src/app/auth/auth.js

```diff
@@ -0,0 +1,152 @@
+angular.module('prx.auth', ['prx.ui.nav'])
+.config(function (PRXDrawerProvider) {
+  PRXDrawerProvider.register({
+    name: 'Sign In',
+    weight: PRXDrawerProvider.BOTTOM,
+    href: 'http://www.prx.org/sessions/new',
+    nav: true,
+    template: '<prx-auth-badge></prx-auth-badge>',
+    dropdownTemplate: '<prx-auth-window></prx-auth-window>'
+  });
+}).service('PrxAuth', function (ngHal, $q, $window) {
+  var PrxAuth = this,
+      deferred = $q.defer(),
+      currentUser = {loggedIn: false};
+  this.$$loginPromise = deferred.promise;
+
+  this.$checkLoggedIn = function () {
+    var iframe = angular.element('<iframe>'),
+        $body = angular.element($window.document.body),
+        uri = [FEAT.ID_SERVER + '/authorize?client_id=' + FEAT.ID_CLIENT_KEY],
+        nonce = [];
+    for (var i=0; i<8; i++) {
+      nonce.push(randomInt(0, 15).toString(16));
+    }
+    nonce = nonce.join('');
+    uri.push("&nonce=", nonce, "&response_type=token&prompt=none");
+    iframe.css('display', 'none');
+    iframe.attr('src', uri.join(''));
+
+    iframe.on('load', function onLoad() {
+      var query = false, accessToken;
+      try {
+        query = iframe[0].contentDocument.location.hash.replace(/^#/,'');
+      } catch (e) {}
+      if (query && (accessToken = parseQuery(query).access_token)) {
+        PrxAuth.$processToken(accessToken).then(function (user) {
+          deferred.resolve(user);
+        });
+      } else {
+        deferred.reject('not logged in');
+      }
+      iframe.off('load', onLoad);
+      iframe.remove();
+      iframe = undefined;
+    });
+
+    $body.append(iframe);
+  };
+
+  this.$processToken = function (token) {
+    this.token = token;
+    this.loggedIn = true;
+    return ngHal.follow('prx:authorization', {}, headers(token)).then(function (result) {
+      angular.copy({
+        loggedIn: true,
+        login: result.name,
+        accountUrl: result.link('prx:default-account').url(),
+        token: token
+      }, currentUser);
+      return result.follow('prx:default-account');
+    }, function (result) {
+      console.error(result);
+      return $q.reject(result);
+    }).then(function (account) {
+      currentUser.account = account;
+      return account.follow('prx:image');
+    }).then(function(image) {
+      currentUser.imageUrl = image.link('enclosure').url();
+      return currentUser;
+    });
+  };
+
+  this.currentUser = function (required) {
+    if (required) {
+      return this.$$loginPromise;
+    } else {
+      return this.$$loginPromise['catch'](function (rejection) {
+        return currentUser;
+      });
+    }
+  };
```

Comment:
Allows us to pull a promise for the currentUser. Couple of things here.

First, if `required`, we reject the promise when not logged in rather than resolving with `{loggedIn: false}`. We do this basically for state `resolve`s, where we want to prevent completing the transition if the user isn't logged in, and for convenience when stacking promises (`PrxAuth.currentUser(true).then(function (user) { return user.account.build('prx:audio-file') })`)

We're using a global promise rather than just returning the `currentUser` object (which, remember, we always have) because we want to ensure that if something is going on that might change the value, we wait until that finishes, and it lets us stack things nicely (remember above) since once the promise is fully resolved we are guaranteed to have all of the properties written to the currentUser object (or none of them).